docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,505 @@
+    };
+
+    VerifyCSharpDiagnostic(test, expected);
+
+    VerifyCSharpFix(test, fixTest);
+}
+```
+
+The preceding code also made a couple changes to the code that builds the expected diagnostic result. It uses the public constants registered in the `MakeConst` analyzer. In addition, it uses two string constants for the input and fixed source:
+
+[!code-csharp[string constants for fix test](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#FirstFixTest "string constants for fix test")]
+
+Run these two tests to make sure they pass. In Visual Studio, open the **Test Explorer** by selecting **Test** > **Windows** > **Test Explorer**.  The press the **Run All** link.
+
+## Create tests for valid declarations
+
+As a general rule, analyzers should exist as quickly as possible, doing minimal work. Visual Studio calls registered analyzers as the user edits code. Responsiveness is a key requirement.  There are a number of tests to add for declarations that should not raise your diagnostic. There are several test cases for code that should not raise your diagnostic. Your analyzer already handles one of those tests, the case where a variable is assigned after being initialized. Add the following string constant to your tests to represent that case:
+
+[!code-csharp[variable assigned](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#VariableAssigned "a variable that is assigned after being initialized won't raise the diagnostic")]
+
+Then, add a data row for this test as shown in the snippet below:
+
+```csharp
+[DataTestMethod]
+[DataRow(""),
+ DataRow(VariableAssigned)]
+public void WhenTestCodeIsValidNoDiagnosticIsTriggered(string testCode)
+```
+
+This test passes as well. Next, add constants for conditions you haven't handled yet:
+
+* Declarations that are already `const`, because they are already const:
+
+   [!code-csharp[already const declaration](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AlreadyConst "a declaration that is already const should not raise the diagnostic")]
+
+* Declarations that have no initializer, because there is no value to use:
+
+   [!code-csharp[declarations that have no initializer](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#NoInitializer "a declaration that has no initializer should not raise the diagnostic")]
+
+* Declarations where the initializer is not a constant, because they can't be compile-time constants:
+
+   [!code-csharp[declarations where the initializer isn't const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#InitializerNotConstant "a declaration where the initializer is not a compile-time constant should not raise the diagnostic")]
+
+All of the preceding conditions are complicated a C# rule that allows multiple declarations as part of one statement. Consider the following test case string constant:
```

Comment:
The first sentence didn't make sense to me.

Reply:
reworded.